### PR TITLE
Propagate the Lookup instance during the Sentry.init() and other initialization workflows.

### PR DESCRIPTION
**Pull request by metlos**
This should make us ready to have a fully customizable and composable configuration.

There are static helpers in `SentryOptions` (e.g. the `defaults()` method) that can help users get up and running quickly but the `SentryOptions` instance can be obtained using a fully configured `Lookup` instance and can also be supplied overrides for both `Dsn` and `SentryClientFactory`.

~~There should be just 1 behavioral change in this PR:~~

~~It is no longer possible to use `SentryOptions.resourceLoader` to inject the resource loader into the default Lookup instance. This needs to be done by manually constructing the `Lookup` instance, using it to construct `SentryOptions` instance and call `Sentry.init()` with that options instance.~~

### CODE REVIEW

**Comment by metlos** [comment]
To address the comment https://github.com/getsentry/sentry-java/pull/759#discussion_r318654151 about why to replace the `AtomicBoolean autoInitAttempted` guard on a volatile `storedClient` field in the `Sentry` class.

The way `storedClient` was mutated was very prone to weird behavior when accessed concurrently.

Take for example the `close()` method. Originally, it looked like this:
```java
    public static void close() {
1         if (storedClient == null) {
2             return;
3         }
4 
5         storedClient.closeConnection();
6         storedClient = null;
7
8         // Allow the client to be auto initialized on the next use.
9         storedClient.closeConnection();
10        autoInitAttempted.set(false);
    }
```

because the `storedClient` is merely a volatile field, the following is easily possible:

`storedClient != null`

| T1 | T2 |
|-----|-----|
| line 5| - |
| -  | line 1|
| line 6| - |
| - | line 5|

at which point we have an NPE in thread T2. 

At this point in time it is easy to eliminate the `autoInitAttempted` boolean, because it actually didn't serve any purpose because it was only used in a synchronized block.

When I think about it, my changes actually introduce a behavioral change. It was possible before to call this sequence:

1. Sentry.getStoredClient()
1. Sentry.setStoredClient(null)
1. Sentry.getStoredClient()

and one would get null from the last call, because `autoInitAttempted` was not reset. In the new impl, the method never returns null (contrary to the javadoc). Not sure how important that change is.

**Comment by bruno-garcia** [comment]
Good point. happy with the full lock.
One question though: What about an atomic reference swap.
Like in: https://github.com/getsentry/sentry-dotnet/blob/1040fe65575a47a3604f406956746500d6a6adcd/src/Sentry/SentrySdk.cs#L96

Just splitting hairs here, happy to merge the lock.

**Comment by metlos** [comment]
Well, as far as I looked and understood the .NET version - there is no concept of autoinit there, am I right? You just use the "disabled instance" until init is called.

The current behavior in the java client is different because `getStoredClient` actually calls init and sets the client on first use. So in the java version - there is the assumption that you get a configured client after the first call, unlike in .NET where, as far as I understood, clients use the disabled instance until init is called.

So we can either change the behavior to match the .NET version or we need to use the locks because there is no operation that would allow for calling a function (the init in our case) during the atomic update.

**Comment by bruno-garcia** [comment]
That makes sense. I wouldn't change the behavior now so happy with the lock.
Thanks for the clarification.

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java [hunk 1]
```diff
@@ -61,8 +63,10 @@ public AndroidSentryClientFactory(Application app) {
      * Construct an AndroidSentryClientFactory using the specified Android Context.
      *
      * @param ctx Android Context.
+     * @param lookup the lookup for locating the configuration
      */
-    public AndroidSentryClientFactory(Context ctx) {
+    public AndroidSentryClientFactory(Context ctx, Lookup lookup) {
```
Comment (bruno-garcia) [inline sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: This would be a breaking change then I guess? We don't offer an overload that doesn't take Lookup where we instantiate it ourselves to be backwards compatible?

File: sentry/src/main/java/io/sentry/Sentry.java [hunk 2]
```diff
@@ -1,37 +1,33 @@
...
-     */
-    private static volatile SentryClient storedClient = null;
-    /**
-     * Tracks whether the {@link #init()} method has already been attempted automatically
-     * by {@link #getStoredClient()}.
+     * A synchronization guard of the stored client. Not using the class as the guard because that could theoretically
+     * deadlock with 3rd party code if it also synced on the class.
      */
-    private static AtomicBoolean autoInitAttempted = new AtomicBoolean(false);
+    private static final Object STORED_CLIENT_ACCESS = new Object();
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/Sentry.java]: It makes sense to get the lock here as you've clarified but is this change needed to get rid of testNG?

File: sentry/src/main/java/io/sentry/Sentry.java [hunk 3]
```diff
@@ -40,25 +36,18 @@ private Sentry() {
...
      * a {@link Dsn} and using the provided {@link SentryClientFactory}.
      *
      * @param sentryClientFactory SentryClientFactory to use.
      * @return SentryClient
+     *
+     * @deprecated use {@link #init(SentryOptions)}
      */
-    public static SentryClient init(SentryClientFactory sentryClientFactory) {
-        return init(null, sentryClientFactory);
+    @Deprecated
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/Sentry.java]: Is it also needed to deprecate the main init function to get rid of testNG? 
Maybe this could be a separate PR so we can discuss?

File: sentry/src/main/java/io/sentry/SentryClientFactory.java [hunk 4]
```diff
@@ -1,80 +1,80 @@
...
-     * @param dsn Data Source Name of the Sentry server.
-     * @param sentryClientFactory SentryClientFactory instance to use, or null to do a config lookup.
-     * @return SentryClient instance, or null if one couldn't be constructed.
+     * @param lookup the lookup instance to use for reading the configuration
+     * @param dsn the DSN instance
+     * @return the Sentry client factory or null if not available
      */
-    public static SentryClient sentryClient(String dsn, SentryClientFactory sentryClientFactory) {
-        Dsn realDsn = resolveDsn(dsn);
+    public static @Nullable SentryClientFactory instantiateFrom(Lookup lookup, Dsn dsn) {
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryClientFactory.java]: We wouldn't want to expose Dsn type anymore. Generally the SDKs did (and .NET for example still does) but we've taken the approach of exposing the String only which has some advantages like easily binded via configuration files and just simpler for developers/code samples etc.

Also I'd suggest we discuss that on a separate PR unrelated to testNG

File: sentry/src/main/java/io/sentry/SentryOptions.java [hunk 5]
```diff
@@ -1,60 +1,311 @@
...
+     * Creates a new instance of the options using the provided parameters.
+     *
+     * @param lookup        the lookup to locate the configuration with
+     * @param dsn           the dsn to use
+     * @param clientFactory the client factory to use
+     * @throws NullPointerException if any of the parameters is null
+     */
+    public SentryOptions(Lookup lookup, Dsn dsn, SentryClientFactory clientFactory) {
+        this.lookup = requireNonNull(lookup, "lookup");
+        this.dsn = requireNonNull(dsn, "dsn");
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryOptions.java]: An empty dsn is a valid use-case. We just keep the SDK disabled. This is a documented behavior we have in all Sentry SDKs.
Also having a public setter for dsn allows users read their dsn after the options was instantiated. Say a framework that creates the options instance for you and binds values from some configuration source you define.

File: sentry/src/main/java/io/sentry/SentryOptions.java [hunk 6]
```diff
@@ -1,60 +1,311 @@
...
+     * @param lookup  the lookup to locate the configuration with
+     * @param dsn     the optional dsn to use or null if the value should be located in the config
+     * @param factory the client factory to use or null if the value should be located in the config
+     * @return new instance of SentryOptions
+     */
+    public static SentryOptions from(Lookup lookup, @Nullable String dsn, @Nullable SentryClientFactory factory) {
+        return from(lookup, resolveDsn(lookup, dsn), factory);
+    }
+
+    private static SentryOptions from(Lookup lookup, Dsn dsn, @Nullable SentryClientFactory factory) {
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryOptions.java]: By not exposing Dsn we get rid of some of these overloads too. Reduce the API surface.

File: sentry/src/main/java/io/sentry/SentryOptions.java [hunk 7]
```diff
@@ -1,60 +1,311 @@
...
+        if (factory == null) {
+            factory = SentryClientFactory.instantiateFrom(lookup, dsn);
+        }
+
+        if (factory == null) {
+            logger.error("Failed to find a Sentry client factory in the provided configuration. Will continue"
+                    + " with a dummy implementation that will send no data.");
+
+            factory = new InvalidSentryClientFactory();
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryOptions.java]: Can we call it NoOpSentryClientFactory or something like that? Invalid seems like I wouldn't want that instance around.

File: sentry/src/main/java/io/sentry/SentryOptions.java [hunk 8]
```diff
@@ -1,60 +1,311 @@
...
+        }
+    }
+
+    // the below is support for instantiating the default lookup
+
+    private static List<ConfigurationResourceLocator> getDefaultResourceLocators() {
+        return asList(new SystemPropertiesBasedLocator(), new EnvironmentBasedLocator(), new StaticFileLocator());
+    }
+
+    private static List<ConfigurationProvider> getDefaultHighPriorityConfigurationProviders() {
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryOptions.java]: The comment above regards POJO applies to these methods too. Would they fit better in some helper?

File: sentry/src/main/java/io/sentry/Sentry.java (hunk 3 above)
Comment (metlos) [reply sentry/src/main/java/io/sentry/Sentry.java]: This method uses the statically initialized lookup instance, which I thought we're trying to shy away from. Therefore I deprecated this method and pointed towards the new way of doing the initialization that supports passing in the configurable sentry options and lookup.

Comment (metlos) [reply sentry/src/main/java/io/sentry/Sentry.java]: We're in a PR that does not deal with TestNG removal (and which I built the TestNG PR on top of), so that's way it seems like mixing 2 things, were the deprecation was just done first.

File: sentry/src/main/java/io/sentry/Sentry.java (hunk 2 above)
Comment (metlos) [reply sentry/src/main/java/io/sentry/Sentry.java]: You are right that this strictly doesn't belong here even if we consider just the propagation of the Lookup instance, which this PR is really about. If you insist, I can revert it back and provide another PR to implement just this change.

Comment (bruno-garcia) [reply sentry/src/main/java/io/sentry/Sentry.java]: It's fine we already clarified why it's needed and what it solves though. Happy to merge in this PR

File: sentry/src/main/java/io/sentry/Sentry.java (hunk 3 above)
Comment (bruno-garcia) [reply sentry/src/main/java/io/sentry/Sentry.java]: If we at least can mock it all while testing, that'd be enough. I would avoid forcing consumers to simply call an overload with `new Lookup()` as that doesn't add a lot of value to them right now. With regards to introducing breaking changes or `@Deprecated` we can take a look at a bigger picture to where this SDK is going to go in the next few months and decide at that level. For now being able to use those tests using normal mocks as opposed to testing those statics is a huge win.

File: sentry/src/main/java/io/sentry/SentryClientFactory.java (hunk 4 above)
Comment (metlos) [reply sentry/src/main/java/io/sentry/SentryClientFactory.java]: Since this is a new method, I think we can safely take the liberty and use `String` as the type for the dsn. That's what I've done in the subsequent PRs, so I think we can resolve this.

Comment (bruno-garcia) [reply sentry/src/main/java/io/sentry/SentryClientFactory.java]: That works

File: sentry/src/main/java/io/sentry/SentryOptions.java (hunk 5 above)
Comment (metlos) [reply sentry/src/main/java/io/sentry/SentryOptions.java]: This constructor now accepts dsn as a nullable string as well as a nullable client factory. Internally, these two fields will never be null, because they will be initialized from the lookup (or use the defaults if not found there at all) if a null value is supplied.

Also I've made the properties of `SentryOptions` mutable again (i.e. make them non-final and provide setters for them), even though I'm not sure that is a great idea - but I don't know the codebase that well to be able to actually judge.

File: sentry/src/main/java/io/sentry/SentryOptions.java (hunk 6 above)
Comment (metlos) [reply sentry/src/main/java/io/sentry/SentryOptions.java]: Removed. The `SentryOptions` now uses string for the dsn.

File: sentry/src/main/java/io/sentry/SentryOptions.java (hunk 7 above)
Comment (metlos) [reply sentry/src/main/java/io/sentry/SentryOptions.java]: I did this to emulate the behavior of `SentryClientFactory#getClient()` which may return `null` but as far as I looked in the codebase such possibility is nowhere taken into account and would result in NPEs.

Because `SentryOptions` doesn't return a client but rather a client factory, I needed to use a factory that would return null client under the same circumstances as the factory. I believe such factory is actually invalid, because it returns null client, which the callers don't expect, hence the naming.

File: sentry/src/main/java/io/sentry/SentryOptions.java (hunk 8 above)
Comment (metlos) [reply sentry/src/main/java/io/sentry/SentryOptions.java]: I moved this back into `Lookup` where it really belongs.

File: sentry/src/main/java/io/sentry/SentryOptions.java (hunk 5 above)
Comment (bruno-garcia) [reply sentry/src/main/java/io/sentry/SentryOptions.java]: Yeah ideally we'd have immutable objects but in this case it's a configuration object useful to separate instantiation from writing values to (i.e binding config files data to it). Someone holding the reference to this instance and changing values at runtime is going out of their way to misuse the API and it's just not something we have to support/worry.
